rectools/visuals/metrics_app.py

```diff
@@ -0,0 +1,194 @@
+import typing as tp
+from functools import lru_cache
+
+import ipywidgets as widgets
+import pandas as pd
+import plotly.graph_objs as go
+from IPython.display import display
+
+from rectools import Columns
+
+WIDGET_WIDTH = 800
+WIDGET_HEIGHT = 500
+TOP_CHART_MARGIN = 20
+LEGEND_TITLE = "Models"
+
+
+class MetricsApp:
+    """
+    Jupyter widgets app for metric visualization and comparison.
+    Do not create instances of this class directly. Use `MetricsApp.construct` method instead.
+    """
+
+    def __init__(
+        self,
+        models_metrics: pd.DataFrame,
+        show_legend: bool = True,
+        auto_display: bool = True,
+        layout_kwargs: tp.Optional[tp.Dict[str, tp.Any]] = None,
+    ):
+        self.models_metrics = models_metrics
+        self.show_legend = show_legend
+        self.auto_display = auto_display
+        self.layout_kwargs = layout_kwargs if layout_kwargs is not None else {}
+        self.fig = go.Figure()
+
+        self._validate_input_data()
+        if self.auto_display:
+            self.display()
+
+    @classmethod
+    def construct(
+        cls,
+        models_metrics: pd.DataFrame,
+        show_legend: bool = True,
+        auto_display: bool = True,
+        layout_kwargs: tp.Optional[tp.Dict[str, tp.Any]] = None,
+    ) -> "MetricsApp":
+        r"""
+        Construct interactive widget for metric-to-metric trade-off analysis.
+
+        Parameters
+        ----------
+        models_metrics : pd.DataFrame
+            A pandas DataFrame containing metrics for visualization. Required columns:
+                - `Columns.Models` - model names
+                - `Columns.Split` - fold number
+                - Any other numeric columns which represent metric values
+        show_legend : bool, default True
+            Specifies whether to display the chart legend.
+        auto_display : bool, default True
+            Automatically displays the widgets immediately after initialization.
+         layout_kwargs : tp.Optional[tp.Dict[str, tp.Any]], optional, default None
+            Additional arguments from `plotly.graph_objects.Layout`
+
+        Returns
+        -------
+        MetricsApp
+            An instance of `MetricsApp`, providing interactive Jupyter widget for metric visualization.
+
+        Examples
+        --------
+        Create interactive widget
```

Comment:
Examples in the documentation are not rendered correctly right now. You can check rendering for your PR here: https://rectools--140.org.readthedocs.build/en/140/api/rectools.visuals.metrics_app.MetricsApp.html
Me guess is that you are missing empty line between text (`Create interactive widget`) and code example. Let's try and add it.

When all is good examples will be rendered just like here: https://rectools--140.org.readthedocs.build/en/140/api/rectools.visuals.visual_app.VisualApp.html

Reply:
Oh, sorry for that. I would add extensions to my VS code to check docs renders of future commits. Fixed